src/main/kotlin/com/trackforever/controllers/ProjectController.kt

```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
+                        currentProject.issues.remove(matchingIssue.id)
+                    } catch (e: UninitializedPropertyAccessException) { // matchingIssue was never initialized
+                        // No need to delete anything.
+                    }
+                    val calculatedHash = generateHash(currentIssue)
+                    if (calculatedHash != currentIssue.hash) {
+                        logger.debug("Mismatched issue hash detected.")
+                        logger.debug("Given hash: ${currentIssue.hash}")
+                        logger.debug("Expected hash: $calculatedHash")
+                        currentIssue.hash = calculatedHash
+                    }
+                    val issuesList = successes[currentIssue.projectId]
+                    if (issuesList != null) {
+                        issuesList += currentIssue
+                    }
+                    currentProject.issues[currentIssue.id] = currentIssue
                 }
+                projectRepository.save(currentProject) // Update the entry in the database
             }
         }
         return when {
-            failedProjects.isEmpty() -> ResponseEntity(HttpStatus.OK)
-            failedProjects.keys.size == content.keys.size -> ResponseEntity(content, HttpStatus.GONE) // None of the specified projects exist.
-            else -> ResponseEntity(failedProjects, HttpStatus.MULTI_STATUS)
+            successes.isEmpty() -> ResponseEntity(HttpStatus.GONE) // None of the projects were found.
+            successes.keys.size == content.keys.size -> ResponseEntity(successes, HttpStatus.OK) // All projects were found
+            else -> ResponseEntity(successes, HttpStatus.MULTI_STATUS)
         }
     }
 
     /**
      * Adds the given projects to the database.
      * Expects a List of projects in a JSON String.
-     * Possible Responses: HTTP 200 OK upon completion.
-     * TODO: Verify that the projects being passed in are valid
+     * Possible Responses: HTTP 200 OK upon completion with a List of projects
+     *                     that were successfully saved.
      */
     @CrossOrigin
     @PutMapping("/projects")
-    fun setProjects(@RequestBody projects: String): ResponseEntity<String> {
+    fun setProjects(@RequestBody projects: String): ResponseEntity<List<TrackForeverProject>> {
         logger.debug("set projects :: ($projects)")
         val mapper = jacksonObjectMapper()
         val content: List<TrackForeverProject> = mapper.readValue(projects)
+        val successes: MutableList<TrackForeverProject> = mutableListOf()
         // Add to the database
         content.forEach {
-            projectRepository.save(it)
+            // Look in the database for the previous project.
+            val serverProject: TrackForeverProject
+            try { // Project with a prevHash found
+                serverProject = projectRepository.findByHash(it.prevHash)
```

Comment:
One slight issue with this is that a project can be updated by another user, then the prevHash will never match and a new project will be created instead. You should probably find by project id and then ignore if the prev hash doesn't match. It's the client's duty to merge in remote changes.